internal/pkg/vm/internal/vmcontext/vmcontex2.go

```diff
@@ -65,15 +66,80 @@ type actorStorage struct {
 }
 
 // NewVM creates a new runtime for executing messages.
-func NewVM(rnd RandomnessSource, actorImpls ActorImplLookup, store storage.VMStorage, st state.Tree) VM {
+func NewVM(rnd RandomnessSource, actorImpls ActorImplLookup, store *storage.VMStorage, st state.Tree) VM {
 	return VM{
-		rnd:        rnd,
-		actorImpls: actorImpls,
-		store:      store,
-		state:      state.NewCachedTree(st),
+		rnd:          rnd,
+		actorImpls:   actorImpls,
+		store:        store,
+		state:        state.NewCachedTree(st),
+		currentEpoch: *types.NewBlockHeight(0),
 	}
 }
 
+// ApplyGenesisMessage forces the execution of a message in the vm actor.
+//
+// This method is intended to be used in the generation of the genesis block only.
+func (vm *VM) ApplyGenesisMessage(from address.Address, to address.Address, method types.MethodID, value types.AttoFIL, params ...interface{}) (interface{}, error) {
+	// get the params into bytes
+	encodedParams, err := abi.ToEncodedValues(params...)
+	if err != nil {
+		return nil, err
+	}
+
+	// normalize from addr
+	from = vm.normalizeFrom(from)
+
+	// build internal message
+	imsg := internalMessage{
+		miner:  address.Undef,
+		from:   from,
+		to:     to,
+		value:  value,
+		method: method,
+		params: encodedParams,
+	}
+
+	ret, err := vm.applyImplicitMessage(imsg)
+	if err != nil {
+		return ret, err
+	}
+
+	// commit state
+	// flush all objects out
+	if err := vm.store.Flush(); err != nil {
+		return nil, err
+	}
+	// commit new actor state
+	if err := vm.state.Commit(context.Background()); err != nil {
+		return nil, err
+	}
+	// TODO: update state root (issue: #3718)
+
+	return ret, nil
+}
+
+func (vm *VM) normalizeFrom(from address.Address) address.Address {
+	// resolve the target address via the InitActor, and attempt to load state.
+	initActorEntry, err := vm.state.GetActor(context.Background(), address.InitAddress)
+	if err != nil {
+		panic(fmt.Errorf("init actor not found. %s", err))
+	}
+
+	// build state handle
+	var stateHandle = NewReadonlyStateHandle(vm.Storage(), initActorEntry.Head)
+
+	// get a view into the actor state
+	initView := initactor.NewView(stateHandle, vm.Storage())
+
+	// lookup the ActorID based on the address
+	targetIDAddr, ok := initView.GetIDAddressByAddress(from)
+	if !ok {
+		panic("TODO")
+	}
+
+	return targetIDAddr
+}
+
 // implement VMInterpreter for VM
 
 var _ interpreter.VMInterpreter = (*VM)(nil)
```

Comment:
Saw this package for the first time, not this PR but note that `makeEpostMessage` and implicit apply is not how we're doing things so you can delete it.

Reply:
we need to do a cleaning pass on all the circular dependencies of deprecated code that makes pieces look alive when they are not needed.